### PR TITLE
Add S.sum and S.product

### PR DESCRIPTION
**Pull request by svozza**
I had to move a few functions to the top of the file so that we could replace the Ramda versions used throughout the code.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2529,11 +2543,17 @@
...
   //. 2
   //. ```
-  S.add =
-  def('add',
-      {},
-      [$.FiniteNumber, $.FiniteNumber, $.FiniteNumber],
-      function(a, b) { return a + b; });
+  S.add = add;
+
+  //# sum :: [FiniteNumber] -> FiniteNumber
```
Comment (davidchambers) [inline index.js]: Should the type be as follows?

```haskell
sum :: Foldable f => f FiniteNumber -> FiniteNumber
```

File: index.js [hunk 2]
```diff
@@ -2591,6 +2611,20 @@
...
+  //.
+  //. ```javascript
+  //. > S.product([1, 2, 3, 4, 5])
+  //. 120
+  //. ```
+  S.product =
+  def('product',
+      {},
+      [$.Array($.FiniteNumber), $.FiniteNumber],
+      reduce(S.mult, 1));
```
Comment (davidchambers) [inline index.js]: `s/S[.]//`

File: index.js [hunk 3]
```diff
@@ -369,6 +369,34 @@
...
+  def('add',
+      {},
+      [$.FiniteNumber, $.FiniteNumber, $.FiniteNumber],
+      function(a, b) { return a + b; });
+
+  var sum =
+  def('sum',
+      {},
+      [$.Array($.FiniteNumber), $.FiniteNumber],
+      reduce(add, 0));
```
Comment (davidchambers) [inline index.js]: I'd be totally okay with replacing `add` with `function(a, b) { return a + b; }` here. I'm not sure it's worth reusing a trivial function if it forces us to separate a function's implementation from its documentation. What do you think?

File: index.js [hunk 4]
```diff
@@ -369,6 +369,34 @@
         }
       });
 
+  var reduce =
```
Comment (davidchambers) [inline index.js]: Let's include a type signature.

File: index.js (hunk 3 above)
Comment (svozza) [inline index.js]: Yep, I agree.

**Comment by svozza** [comment]
Just realised I forgot to do `git add` with the tests!

File: index.js [hunk 5]
```diff
@@ -369,6 +369,30 @@
...
+        }
+      });
+
+  //  sum :: Foldable f => f FiniteNumber -> FiniteNumber
+  var sum =
+  def('sum',
+      {},
+      [$.Array($.FiniteNumber), $.FiniteNumber],
+      reduce(function(a, b) { return a + b; }, 0));
+
```
Comment (davidchambers) [inline index.js]: It looks as though each of these functions can now be moved alongside its description. `sum` is referenced in the definition of `meld`, but will have been defined by the time the body of that function is executed.

File: index.js [hunk 6]
```diff
@@ -2535,6 +2545,16 @@
       [$.FiniteNumber, $.FiniteNumber, $.FiniteNumber],
       function(a, b) { return a + b; });
 
+  //#  sum :: Foldable f => f FiniteNumber -> FiniteNumber
+  //.
+  //. Returns the sum of a list of (finite) numbers.
```
Comment (davidchambers) [inline index.js]: In light of #180, let's `s/a list/an array/`.

Also, `sum = K(0)` would fit with this description, as it always returns the sum of `[]`, an array of finite numbers. ;)

Let's write *the array* rather than *an array*. I suggest:

_Returns the sum of the given array of (finite) numbers._

File: index.js [hunk 7]
```diff
@@ -2535,6 +2545,16 @@
       [$.FiniteNumber, $.FiniteNumber, $.FiniteNumber],
       function(a, b) { return a + b; });
 
+  //#  sum :: Foldable f => f FiniteNumber -> FiniteNumber
```
Comment (davidchambers) [inline index.js]: `s/# /#/`

File: test/product.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                   'product :: Array FiniteNumber -> FiniteNumber\n' +
+                   '                 ^^^^^^^^^^^^\n' +
+                   '                      1\n' +
+                   '\n' +
+                   '1)  -Infinity :: Number, ValidNumber\n' +
+                   '\n' +
+                   'The value at position 1 is not a member of ‘FiniteNumber’.\n'));
+  });
+
+  it('returns the product of a list of numbers', function() {
```
Comment (davidchambers) [inline test/product.js]: `s/a list/an array/`

File: test/product.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    eq(S.product([1, 2, 3, 4, 5]), 120);
+    eq(S.product([1, 2, 3, 4, -5]), -120);
+  });
```
Comment (davidchambers) [inline test/product.js]: Let's also test `[]`, an array containing `0`, and an array containing `-0`.

File: test/sum.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+                   '\n' +
+                   '1)  -Infinity :: Number, ValidNumber\n' +
+                   '\n' +
+                   'The value at position 1 is not a member of ‘FiniteNumber’.\n'));
+  });
+
+  it('sums a list of numbers together', function() {
+    eq(S.sum([1, 2, 3, 4, 5]), 15);
+    eq(S.sum([1, 2, 3, 4, -5]), 5);
+  });
```
Comment (davidchambers) [inline test/sum.js]: Let's also test `[]`.

File: index.js [hunk 11]
```diff
@@ -2585,12 +2605,26 @@
...
+  //. Returns the product of a list (finite) numbers.
+  //.
+  //. ```javascript
+  //. > S.product([1, 2, 3, 4, 5])
+  //. 120
+  //. ```
+  S.product =
+  def('product',
+      {},
+      [$.Array($.FiniteNumber), $.FiniteNumber],
```
Comment (davidchambers) [inline index.js]: I think the best we can currently do here is `{f: [Foldable]}, [f, $.FiniteNumber]`.

File: index.js (hunk 5 above)
Comment (svozza) [inline index.js]: No, the tests fail if I move them and do something like 

```js
var sum = S.sum =
```

**Comment by davidchambers** [comment]
Let's also test these:

```javascript
S.sum(S.Nothing())
S.sum(S.Just(42))
S.sum(S.Left('XXX'))
S.sum(S.Right(42))
```

The last two won't work until we resolve #119, so we could put these assertions in `it.skip` blocks.

File: index.js (hunk 5 above)
Comment (davidchambers) [inline index.js]: Ah, because `meld` is applied in the definition of `mapMaybe`. Let's change that definition from `meld([R.map, justs])` to something along these lines:

```javascript
function(f, xs) { return justs(R.map(f, xs)); }
```

File: index.js (hunk 11 above)
Comment (svozza) [inline index.js]: Interesting, I found a bug in the `Foldable` type check function. I think we never saw it before because it always short-circuited on the array check.

**Comment by svozza** [comment]
I've updated it to use the Foldable type constraint. I'm not happy with the error messages for the `Infinity` cases. The `product` message now leaks that it is implemented in terms of `mult` and `sum` gives an error based on the return value so it actually allows the disallowed computation occur.

**Comment by davidchambers** [comment]
> The `product` message now leaks that it is implemented in terms of `mult` and `sum` gives an error based on the return value so it actually allows the disallowed computation occur.

I suggest replacing `mult` with `function(a, b) { return a * b; }` so we refer to the right function.

Complaining about the return value isn't ideal, but it's the best we can do until sanctuary-def permits us to express `Foldable f => f FiniteNumber`.

File: index.js (hunk 5 above)
Comment (svozza) [inline index.js]: :zap:

**Comment by svozza** [comment]
Yeah, it's not great but I guess there's nothing we can do.

File: index.js [hunk 12]
```diff
@@ -2378,21 +2396,7 @@
...
-          var acc = initial;
-          for (var idx = 0; idx < foldable.length; idx += 1) {
-            acc = f(acc, foldable[idx]);
-          }
-          return acc;
-        } else {
-          return foldable.reduce(f, initial);
-        }
-      });
+  S.reduce = reduce;
```
Comment (davidchambers) [inline index.js]: We can revert this change, I believe.

File: index.js [hunk 13]
```diff
@@ -2535,6 +2539,20 @@
...
       function(a, b) { return a + b; });
 
+  //# sum :: Foldable f => f FiniteNumber -> FiniteNumber
+  //.
+  //. Returns the sum of the given array of (finite) numbers.
+  //.
+  //. ```javascript
+  //. > S.sum([1, 2, 3, 4, 5])
+  //. 15
+  //. ```
```
Comment (davidchambers) [inline index.js]: Let's include `S.sum([])`.

File: index.js [hunk 14]
```diff
@@ -2591,6 +2609,20 @@
...
       function(a, b) { return a * b; });
 
+  //# product :: Foldable f => f FiniteNumber -> FiniteNumber
+  //.
+  //. Returns the product of the given array of (finite) numbers.
+  //.
+  //. ```javascript
+  //. > S.product([1, 2, 3, 4, 5])
+  //. 120
+  //. ```
```
Comment (davidchambers) [inline index.js]: Let's include `S.product([])`.

File: index.js (hunk 13 above)
Comment (davidchambers) [inline index.js]: Let's also include `S.sum(S.Nothing())` and `S.sum(S.Just(42))`.

File: index.js (hunk 14 above)
Comment (davidchambers) [inline index.js]: Let's also include `S.product(S.Nothing())` and `S.product(S.Just(42))`.

File: index.js (hunk 12 above)
Comment (svozza) [inline index.js]: :zap:

File: index.js (hunk 13 above)
Comment (svozza) [inline index.js]: :zap:

File: index.js (hunk 14 above)
Comment (svozza) [inline index.js]: :zap:

File: index.js [hunk 15]
```diff
@@ -2378,7 +2379,8 @@
   //. > S.reduce((xs, x) => [x].concat(xs), [], [1, 2, 3, 4, 5])
   //. [5, 4, 3, 2, 1]
   //. ```
-  S.reduce =
+  //  reduce :: Foldable f => (a -> b -> a) -> a -> f b -> a
```
Comment (davidchambers) [inline index.js]: This line is now redundant.

Comment (svozza) [inline index.js]: Oops!

**Comment by svozza** [comment]
Anything else left on this?

**Comment by davidchambers** [comment]
> Anything else left on this?

This looks great, Stefano. This has motivated me to resolve #119. I plan to do so tomorrow. Once that pull request is merged, you'll be able to rebase this one and remove the `.skip`s. :)

File: test/product.js [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+  it('can be applied to Maybes', function() {
+    eq(S.product(S.Nothing()), 1);
+    eq(S.product(S.Just(42)), 42);
+  });
+
+  it.skip('can be applied to Eithers', function() {
```
Comment (davidchambers) [inline test/product.js]: You can remove the `.skip` now thanks to #204. :)

Comment (svozza) [inline test/product.js]: :zap:

File: index.js [hunk 17]
```diff
@@ -2699,6 +2700,29 @@
...
+  //. 42
+  //.
+  //. > S.sum(S.Nothing())
+  //. 0
+  //. ```
+  var sum = S.sum =
+  def('sum',
+      {f: [Foldable]},
+      [f, $.FiniteNumber],
+      reduce_(function(a, b) { return a + b; }, 0));
```
Comment (davidchambers) [inline index.js]: Might this be better?

```javascript
reduce(function(a) { return function(b) { return a + b; }; }, 0)
```

`reduce_` will curry the provided function, but we could remove a step by defining a curried function.

We needn't worry about this, though. Once we can express `f FiniteNumber` in the type system we'll be able to use `S.add` in this definition without references to that function appearing in error messages. :)

Comment (svozza) [inline index.js]: :zap:

**Comment by davidchambers** [comment]
:deciduous_tree: Lovely patch, Stefano!
